scanpy/tools/_ingest.py

```diff
@@ -0,0 +1,239 @@
+
+    def __init__(self, adata):
+        #assume rep is X if all initializations fail to identify it
+        self._rep = adata.X
+        self._use_rep = 'X'
+
+        self._n_pcs = None
+
+        self._adata_ref = adata
+        self._adata_new = None
+
+        if 'pca' in adata.uns:
+            self._init_pca(adata)
+
+        if 'neighbors' in adata.uns:
+            self._init_neighbors(adata)
+
+        if 'X_umap' in adata.obsm:
+            self._init_umap(adata)
+
+        self._obsm = None
+        self._obs = None
+        self._labels = None
+
+        self._indices = None
+        self._distances = None
+
+    def _pca(self, n_pcs=None):
+        X = self._adata_new.X
+        X = X.toarray() if issparse(X) else X.copy()
+        if self._pca_use_hvg:
+            X = X[:, self._adata_ref.var['highly_variable']]
+        if self._pca_centered:
+            X -= X.mean(axis=0)
+        X_pca = np.dot(X, self._pca_basis[:, :n_pcs])
+        return X_pca
+
+    def _same_rep(self):
+        adata = self._adata_new
+        if self._n_pcs is not None:
+            return self._pca(self._n_pcs)
+        if self._use_rep == 'X':
+            return adata.X
+        if self._use_rep in adata.obsm.keys():
+            return adata.obsm[self._use_rep]
+        return adata.X
+
+    def transform(self, adata_new):
+        self._obs = pd.DataFrame(index=adata_new.obs.index)
+        #not sure if it should be AxisArrays
+        self._obsm = AxisArrays(adata_new, 0)
+
+        self._adata_new = adata_new
+        self._obsm['rep'] = self._same_rep()
+
+    def neighbors(self, k=10, queue_size=5, random_state=0):
+        from umap.nndescent import initialise_search
+        from umap.utils import deheap_sort
+        from umap.umap_ import INT32_MAX, INT32_MIN
+
+        random_state = check_random_state(random_state)
+        rng_state = random_state.randint(INT32_MIN, INT32_MAX, 3).astype(np.int64)
+
+        train = self._rep
+        test = self._obsm['rep']
+
+        init = initialise_search(self._rp_forest, train, test, int(k * queue_size),
+                                 self._random_init, self._tree_init, rng_state)
+
+        result = self._search(train, self._search_graph.indptr, self._search_graph.indices, init, test)
+        indices, dists = deheap_sort(result)
+        self._indices, self._distances = indices[:, :k], dists[:, :k]
+
+    def _umap_transform(self):
+        return self._umap.transform(self._obsm['rep'])
+
+    def map_embedding(self, method):
+        if method == 'umap':
+            self._obsm['X_umap'] = self._umap_transform()
+        elif method == 'pca':
+            self._obsm['X_pca'] = self._pca()
+        else:
+            raise NotImplementedError('Ingest supports only umap embeddings for now.')
+
+    def _knn_classify(self, labels):
+        cat_array = self._adata_ref.obs[labels]
+
+        values = [cat_array[inds].mode()[0] for inds in self._indices]
```

Comment:
Would you be interested in adding weighted knn classification? I think all we would need to do is calculate the connectivities, which I think we could do with [these lines](https://github.com/lmcinnes/umap/blob/424268f6a99a401e3b9eae9c1e8dadebf619db83/umap/umap_.py#L1828-L1837).

Reply:
Do you mean distance weighted knn?

Reply:
Something like that, I was thinking UMAP’s connectivity.